### PR TITLE
Incoming controller

### PR DESCRIPTION
**Pull request by tnum**
Incoming controller working for topics (from email subject - if the topic does not exist it is created) and URL (from email body). Still to do: handle unregistered users in new branch.

### CODE REVIEW

File: app/controllers/incoming_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class IncomingController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: [:create]
+
+  def create
+    puts "INCOMING PARAMS HERE: #{params}"
```
Comment (eliotsykes) [inline app/controllers/incoming_controller.rb]: Can be removed if you're finished with this debugging.

File: app/controllers/incoming_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    @user = User.find_by(email: params[:sender])
+
+    if @user.present?
+      @topic = Topic.find_or_create_by(user_id: @user.id, title: params[:subject])
```
Comment (eliotsykes) [inline app/controllers/incoming_controller.rb]: No need to use `_id`-suffix or call `.id`. Can be simplified to:

``` ruby
@topic = Topic.find_or_create_by(user: @user, title: params[:subject])
```

File: app/controllers/incoming_controller.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      @topic.save
```
Comment (eliotsykes) [inline app/controllers/incoming_controller.rb]: Line 11 `@topic.save` can be safely removed. Line 10 will `create` if needed, which does the save to the database.

File: app/controllers/incoming_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      @url = params["body-plain"]
+      @bookmark = Bookmark.create(topic: @topic, url: @url)
+      @bookmark.save
```
Comment (eliotsykes) [inline app/controllers/incoming_controller.rb]: Line 14 can be removed, no need to do a `save` immediately after a `create`.

File: app/controllers/incoming_controller.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    end
+
+    head 200
```
Comment (eliotsykes) [inline app/controllers/incoming_controller.rb]: No need to use `@instance_variables` in this controller action. We can know this as the purpose of `@instance_variables` **in controller actions** is to make them available to the view template. This controller action has no associated view template.

Consider switching the instance variables to local variables, e.g.

``` ruby
@bookmark = Bookmark.create(...)
# becomes:
bookmark = Bookmark.create(...)
```

As a guideline, its normally a good idea to minimise the exposure of variables (this aids certain memory optimisations at runtime **and** reduces the likelihood for accidental variable name collisions). 

Local variables are less exposed than instance variables.

File: config/routes.rb [hunk 6]
```diff
@@ -5,6 +5,8 @@
   	resources :bookmarks
   end
 
+  post :incoming, to: 'incoming#create'
```
Comment (eliotsykes) [inline config/routes.rb:8]: Did this fix the 404 issue?

File: app/controllers/incoming_controller.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    @user = User.find_by(email: params[:sender])
+
+    if @user.present?
+      @topic = Topic.find_or_create_by(user_id: @user.id, title: params[:subject])
+      @topic.save
+      @url = params["body-plain"]
+      @bookmark = Bookmark.create(topic: @topic, url: @url)
+      @bookmark.save
+    end
+
```
Comment (eliotsykes) [inline app/controllers/incoming_controller.rb:13]: Line 17: `head 200` would be more clearly expressed as:

``` ruby
  head :ok
```

All of the possible http status code `:symbols` Rails recognises are listed in the table here: http://guides.rubyonrails.org/layouts_and_rendering.html#the-status-option

File: config/routes.rb (hunk 6 above)
Comment (tnum) [inline config/routes.rb:8]: I'm not sure what was causing the 404 issue, but a while later it resent and worked. I had added that from the tutorial so I think it was something wrong with the incoming_controller.rb

**Comment by tnum** [comment]
Tested and working

**Comment by eliotsykes** [comment]
Excellent :+1:

**Pull request merged**